Common/include/geometry/dual_grid/CEdge.hpp

```diff
@@ -25,154 +24,187 @@
+                      const su2double* coord_Edge_CG,
+                      const su2double* coord_FaceElem_CG,
+                      const su2double* coord_Elem_CG);
 
   /*!
-   * \overload
-   * \brief Set the face that correspond to an edge.
-   * \param[in] val_coord_Edge_CG - Coordinates of the centre of gravity of the edge.
-   * \param[in] val_coord_Elem_CG - Coordinates of the centre of gravity of the element.
+   * \brief Set the face that corresponds to an edge (2D version).
+   * \param[in] iEdge - Edge index.
+   * \param[in] coord_Edge_CG - Coordinates of the centre of gravity of the edge.
+   * \param[in] coord_Elem_CG - Coordinates of the centre of gravity of the element.
    * \param[in] config - Definition of the particular problem.
    * \return Compute the normal (dimensional) to the face that makes the contorl volume boundaries.
    */
-  void SetNodes_Coord(su2double *val_coord_Edge_CG, su2double *val_coord_Elem_CG) override;
+  void SetNodes_Coord(unsigned long iEdge,
+                      const su2double* coord_Edge_CG,
+                      const su2double* coord_Elem_CG);
 
   /*!
    * \brief Copy the the normal vector of a face.
-   * \param[in] val_normal - Vector where the subroutine is goint to copy the normal (dimensional).
+   * \param[in] iEdge - Edge index.
+   * \param[out] normal - Object into which the normal (dimensional) will be copied.
    */
-  inline void GetNormal(su2double *val_normal) const override {
-    for (unsigned short iDim = 0; iDim < nDim; iDim++)
-      val_normal[iDim] = Normal[iDim];
+  template<class T>
+  inline void GetNormal(unsigned long iEdge, T& normal) const {
+    for (auto iDim = 0ul; iDim < Normal.cols(); iDim++)
+      normal[iDim] = Normal(iEdge,iDim);
   }
 
   /*!
    * \brief Get the normal to a face of the control volume asociated with an edge.
+   * \param[in] iEdge - Edge index.
    * \return Dimensional normal vector, the modulus is the area of the face.
    */
-  inline su2double *GetNormal(void) override {  return Normal; }
+  inline const su2double* GetNormal(unsigned long iEdge) const { return Normal[iEdge]; }
 
   /*!
-   * \brief Initialize normal vector.
+   * \brief Initialize normal vector to 0.
    */
-  inline void SetZeroValues(void) override {
-    for (unsigned short iDim = 0; iDim < nDim; iDim ++)
-      Normal[iDim] = 0.0;
-  }
+  void SetZeroValues(void);
 
   /*!
-   * \brief Set the normal vector.
-   * \param[in] val_face_normal - Vector to initialize the normal vector.
+   * \brief Set the normal vector of an edge.
+   * \param[in] iEdge - Edge index.
+   * \param[in] normal - Vector to initialize the normal vector.
    * \return Value of the normal vector.
    */
-  inline void SetNormal(const su2double *val_face_normal) override {
-    for (unsigned short iDim = 0; iDim < nDim; iDim++)
-      Normal[iDim]=val_face_normal[iDim];
+  template<class T>
+  void SetNormal(unsigned long iEdge, const T& normal) {
```

Comment:
Why are they templated? To pass different array types? :+1:

Reply:
That was the idea yes, in case we need to pass `std::vector` or `std::array` or even just an object with `operator []`